### ISSUE TITLE
SC2016 false positive with ripgrep capture groups

### ISSUE THREAD

**Issue by SafariMonkey** [koalaman/shellcheck#2209]
#### For bugs
- Rule Id (if any, e.g. SC1000): SC2016
- My shellcheck version (`shellcheck --version` or 'online'): online
- [x] I tried on shellcheck.net and verified that this is still a problem on the latest commit

#### Here's a snippet or screenshot that shows the problem:

```sh
#!/bin/bash
rg '(.) (.)' -r '$2 $1'
```

#### Here's what shellcheck currently says:
```
Line 2	SC2016: Expressions don't expand in single quotes, use double quotes for that.
```
#### Here's what I wanted or expected to see:

No error, as the `-r` flag to [`rg`](https://github.com/BurntSushi/ripgrep) takes dollar-prefixed capture groups (described e.g. at the bottom of [this section](https://github.com/BurntSushi/ripgrep/blob/master/GUIDE.md#replacements)) and as such, dollar prefixes are not unexpected. (I would guess that fixing this is a matter of adding it somewhere in [this list](https://github.com/koalaman/shellcheck/blob/5669eb22037980c5a6b74b0d420cb452990bcf88/src/ShellCheck/Analytics.hs#L1004-L1028).)

**Comment by TinCanTech** [comment]
`ripgrep` is not installed by default but it is listed in `ubunutu-20` `apt` .. so, one for @koalaman to decide.

**Comment by SafariMonkey** [comment]
As far as I am aware, `sed` and `grep` use backslashes to denote capture groups, while `rg` and `awk` use dollar signs. `sed` does use dollar sign on commands to apply them to the last line, as I understand it, and that is handled elsewhere in that function.

On the other point, I assumed that with tools like `jq` on the list, a popular `grep` alternative (with [significant performance and/or correctness advantages over alternatives](https://blog.burntsushi.net/ripgrep/), not just nicer ergonomics) seemed like it would be a reasonable candidate for inclusion in that list. Of course, if it's not deemed appropriate, I understand.

As an aside, I noticed that `jq` has a comment on it saying that one `could also check that user provides --arg`. This doesn't seem correct to me, as `jq` supports binding variables inline using the `.something as $var` syntax, which I use frequently.

**Comment by TinCanTech** [comment]
> As far as I am aware, `sed` and `grep` use backslashes to denote capture groups

However, `SC2016: Expressions don't expand in single quotes ..` is not about _capture groups_.
`shellcheck` is simply informing you that it noticed shell variables `$1` and `$2` in single quotes `'` 
which do not expand to values.

I've had this conversation before when I thought the same as you.  `shellcheck` throws the same warning when a `grep` or `sed` expression contains characters which would be expanded by the shell if they were not in single quotes.  My question was then: _Why is `shellcheck` checking statements for `grep` and `sed` which are not shell expressions_ ?

> Also, `awk` is assigned as `isProbablyOk` but not `grep` or `sed` ....

My point being here that an expression for `awk` does **not** throw the same warning ..

Hence: _It is difficult to guess/predict exactly what `shellcheck` will spit out sometimes_.. 
> hard to _guess_ the _standard_ here

But `shellcheck` has saved me many hours of debugging, so I'm happy to work with it as it stands.

**Comment by pabs3** [comment]
I too have seen the false positives when shellcheck detects a `$` character in `sed` commands, either the `$` match last line command, or the `$` regex match end of pattern space character.

Probably shellcheck needs a generalised way to detect which arguments of which commands could contain shell metacharacters and ignore those.

**Comment by SafariMonkey** [comment]
Thanks for adding it!

It seems I assumed the list was supposed to be more comprehensive than you intend it to be. I will take your advice and add it to my `.shellcheckrc` if it bothers me in the future. I do think it's a useful lint for if you e.g. add a variable to a string you don't notice is single quoted, but working syntax highlighting should (not) highlight that also, so disabling it shouldn't be too much of a loss.
